src/main/java/org/cloudfoundry/promregator/auth/TokenFetcher.java

```diff
@@ -0,0 +1,170 @@
+
+	protected final CloseableHttpClient httpClient;
+
+	protected TokenFetcher(OAuth2XSUAAAuthenticationConfiguration authConfig)
+			throws GeneralSecurityException, IOException {
+		this.authConfig = authConfig;
+		this.httpClient = prepareHttpCient(authConfig);
+	}
+
+	public abstract TokenResponse getJWT(RequestConfig config) throws IOException;
+
+	protected abstract CloseableHttpClient prepareHttpCient(OAuth2XSUAAAuthenticationConfiguration authConfig)
+			throws GeneralSecurityException, IOException;
+
+	protected HttpPost preparePostRequest(String tokenServiceUrl, RequestConfig config,
+			List<NameValuePair> additionalParams) {
+		HttpPost httpPost = new HttpPost(this.authConfig.getTokenServiceURL());
+		httpPost.setConfig(config);
+
+		/*
+		 * closing the connection afterwards is important! Background: httpclient will
+		 * otherwise try to keep the connection open. We won't be calling often anyway,
+		 * so the server would be drained from resources. Moreover, if the server has
+		 * gone away in the meantime, the next attempt to call would fail with a recv
+		 * error when reading from the socket.
+		 */
+		httpPost.setHeader("Connection", "close");
+
+		List<NameValuePair> params = new ArrayList<NameValuePair>();
+		params.addAll(additionalParams);
+		params.add(new BasicNameValuePair("grant_type", "client_credentials"));
+		if (this.authConfig.getScopes() != null) {
+			params.add(new BasicNameValuePair("scope", this.authConfig.getScopes()));
+		}
+		UrlEncodedFormEntity formEntity = new UrlEncodedFormEntity(params, Consts.UTF_8);
+		httpPost.setEntity(formEntity);
+
+		return httpPost;
+	}
+
+	protected TokenResponse fetchAndParseToken(HttpPost httpPost) throws IOException {
+		try (CloseableHttpResponse response = httpClient.execute(httpPost)) {
+
+			if (response.getStatusLine().getStatusCode() != 200) {
+				log.error(String.format(
+						"Server did not respond with ok while fetching JWT from token server; status code provided: %d",
+						response.getStatusLine().getStatusCode()));
+				return null;
+			}
+
+			return parseToken(response.getEntity());
+
+		} catch (ClientProtocolException e) {
+			log.error("Unable to read from the token server", e);
+			throw e;
+		} catch (IOException e) {
+			log.error("IO Exception while reading from the token server", e);
+			throw e;
+		}
+
+	}
+
+	public void close() throws IOException {
+		httpClient.close();
+	}
+
+	private static TokenResponse parseToken(HttpEntity responseBody) throws IOException {
+
+		try {
+			String json = EntityUtils.toString(responseBody, StandardCharsets.UTF_8);
+
+			if (json == null) {
+				log.warn("Null-JSON detected on OAuth response");
+				return null;
+			}
+
+			return new Gson().fromJson(json, TokenResponse.class);
+
+		} catch (IOException e) {
+			log.error("IO Exception while running GSON parser", e);
+			throw e;
+
+		} catch (ParseException e) {
+			log.error("GSON parser exception on JWT response from token server", e);
+			throw e;
+		}
```

Comment:
The response by the authorization server contains sensitive data. Therefore I would not log the error message of the JSON parser, because it can contain the data in clear text.

Reply:
Obsolet. This code has been removed.